apis/Google.Cloud.Diagnostics.AspNetCore/Google.Cloud.Diagnostics.AspNetCore.IntegrationTests/Trace/TraceTest.cs

```diff
@@ -0,0 +1,420 @@
+                
+                var header = TraceHeaderContext.Create(traceId, spanId, true);
+                client.DefaultRequestHeaders.Add(TraceHeaderContext.TraceHeader, header.ToString());
+                await client.GetAsync($"/Trace/Trace/{testId}");
+            }
+
+            var spanName = TraceController.GetMessage(nameof(TraceController.Trace), testId);
+            var trace = await GetTrace(spanName, startTime);
+
+            Assert.NotNull(trace);
+            Assert.Equal(traceId, trace.TraceId);
+            Assert.Equal(2, trace.Spans.Count);
+            var span = trace.Spans.First(s => s.Name.StartsWith("/Trace"));
+            Assert.Equal(spanId, span.ParentSpanId);
+        }
+
+        [Fact]
+        public async Task Trace_QPS()
+        {
+            string testId = Utils.GetTestId();
+            var startTime = Timestamp.FromDateTime(DateTime.UtcNow);
+
+            var builder = new WebHostBuilder().UseStartup<TraceTestNoBufferApplication>();
+            using (TestServer server = new TestServer(builder))
+            {
+                var client = server.CreateClient();
+
+                // The first request is not traced, sleep long enough 
+                // for the rate limiter to allow a request, then the second
+                // request is traced.
+                await client.GetAsync($"/Trace/TraceLabels/{testId}");
+                Thread.Sleep(TimeSpan.FromSeconds(2));
+                await client.GetAsync($"/Trace/Trace/{testId}");                
+            }
+
+            var spanNameTrace = TraceController.GetMessage(nameof(TraceController.Trace), testId);
+            var spanNameLabels = TraceController.GetMessage(nameof(TraceController.TraceLabels), testId);
+            Assert.Null(await GetTrace(spanNameLabels, startTime, expectTrace: false));
+            Assert.NotNull(await GetTrace(spanNameTrace, startTime));
+        }
+
+        [Fact]
+        public async Task Trace_Buffer()
+        {
+            string testId = Utils.GetTestId();
+            var startTime = Timestamp.FromDateTime(DateTime.UtcNow);
+
+            var builder = new WebHostBuilder().UseStartup<TraceTestBufferHighQpsApplication>();
+            using (TestServer server = new TestServer(builder))
+            {
+                var client = server.CreateClient();
+
+                // Make a trace with a small span that will not cause the buffer to flush.
+                await client.GetAsync($"/Trace/Trace/{testId}");
+                var spanName = TraceController.GetMessage(nameof(TraceController.Trace), testId);
+                Assert.Null(await GetTrace(spanName, startTime, expectTrace: false));
+
+                // Make a large trace that will flush the buffer.
+                await client.GetAsync($"/Trace/TraceStackTrace/{testId}");
+            }
+
+            var spanNameTrace = TraceController.GetMessage(nameof(TraceController.Trace), testId);
```

Comment:
This creates the same name that's stored in `spanName` above. I think it would be better to create it once above the using statement. Then it's clear that getting the trace above returns null but getting the same one below returns an instance.